prow/pod-utils/decorate/podspec.go

```diff
@@ -172,6 +168,177 @@ func ProwJobToPod(pj kube.ProwJob, buildID string) (*v1.Pod, error) {
+func sshVolume(secret string) (kube.Volume, kube.VolumeMount) {
+	var sshKeyMode int32 = 0400 // this is octal, so symbolic ref is `u+r`
+	name := strings.Join([]string{"ssh-keys", secret}, "-")
+	mountPath := path.Join("/secrets/ssh", secret)
+	v := kube.Volume{
+		Name: name,
+		VolumeSource: kube.VolumeSource{
+			Secret: &kube.SecretSource{
+				SecretName:  secret,
+				DefaultMode: &sshKeyMode,
+			},
+		},
+	}
+
+	vm := kube.VolumeMount{
+		Name:      name,
+		MountPath: mountPath,
+		ReadOnly:  true,
+	}
+
+	return v, vm
+}
+
+// cookiefileVolumes converts a secret holding cookies into the corresponding volume and mount.
+//
+// Secret can be of the form secret-name/base-name or just secret-name.
+// Here secret-name refers to the kubernetes secret volume to mount, and base-name refers to the key in the secret
+// where the cookies are stored. The secret-name pattern is equivalent to secret-name/secret-name.
+//
+// This is used by CloneRefs to attach the mount to the clonerefs container.
+// The returned string value is the path to the cookiefile for use with --cookiefile.
+func cookiefileVolume(secret string) (kube.Volume, kube.VolumeMount, string) {
+	// Separate secret-name/key-in-secret
+	parts := strings.SplitN(secret, "/", 2)
+	cookieSecret := parts[0]
+	var base string
+	if len(parts) == 1 {
+		base = parts[0] // Assume key-in-secret == secret-name
+	} else {
+		base = parts[1]
+	}
+	var cookiefileMode int32 = 0400 // u+r
+	vol := kube.Volume{
+		Name: "cookiefile",
+		VolumeSource: kube.VolumeSource{
+			Secret: &kube.SecretSource{
+				SecretName:  cookieSecret,
+				DefaultMode: &cookiefileMode,
+			},
+		},
+	}
+	mount := kube.VolumeMount{
+		Name:      vol.Name,
+		MountPath: "/secrets/cookiefile", // append base to flag
+		ReadOnly:  true,
+	}
+	return vol, mount, path.Join(mount.MountPath, base)
+}
+
+func pointerRefs(refs []kube.Refs) []*kube.Refs {
+	var out []*kube.Refs
+	for _, r := range refs {
+		rp := r // without this every element == &r
+		out = append(out, &rp)
+	}
+	return out
+}
+
+// CloneRefs constructs the container and volumes necessary to clone the refs requested by the ProwJob.
+//
+// The container checks out repositories specified by the ProwJob Refs to `codeMount`.
+// A log of what it checked out is written to `clone.json` in `logMount`.
+//
+// The container may need to mount SSH keys and/or cookiefiles in order to access private refs.
+// CloneRefs returns a list of volumes containing these secrets required by the container.
+func CloneRefs(pj kube.ProwJob, codeMount, logMount kube.VolumeMount) (*kube.Container, []kube.Refs, []kube.Volume, error) {
+	if pj.Spec.DecorationConfig == nil {
+		pj.Spec.DecorationConfig = &kube.DecorationConfig{}
+	}
+	if pj.Spec.DecorationConfig.SkipCloning {
+		return nil, nil, nil, nil
+	}
+	var cloneVolumes []kube.Volume
+	var refs []kube.Refs // Do not return []*kube.Refs which we do not own
```

Comment:
nit: using `[]*kube.Refs` here would eliminate the need for the `pointerRefs` function and make all the ref slices have a consistent type. The elements can be copied as you add them to `refs` to avoid returning pointers that "we do not own".